src/main/java/mate/academy/AsyncRequestProcessor.java

```diff
@@ -1,16 +1,36 @@
 package mate.academy;
 
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 
 public class AsyncRequestProcessor {
+    private static final int MILLIS = 300;
+
     private final Executor executor;
+    private final Map<String, UserData> cache = new ConcurrentHashMap<>();
 
     public AsyncRequestProcessor(Executor executor) {
         this.executor = executor;
     }
 
     public CompletableFuture<UserData> processRequest(String userId) {
-        return null;
+        if (cache.containsKey(userId)) {
+            return CompletableFuture.completedFuture(cache.get(userId));
+        }
+        return CompletableFuture.supplyAsync(() -> getUserDetails(userId), executor);
+    }
+
+    private UserData getUserDetails(String userId) {
+        try {
+            Thread.sleep(MILLIS);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Failed to retrieve details for userId: " + userId, e);
```

Comment:
Wrapping the InterruptedException into a RuntimeException can be problematic because it makes it harder to deal with the interrupted state of the thread. It's usually better to restore the interrupted status by calling `Thread.currentThread().interrupt()` and then either rethrow the exception or return from the method.

Reply:
I don't think it's necessary to interrupt the thread in this case.